src/aws_xray_sdk_clj/impl.clj

```diff
@@ -32,78 +37,127 @@
+  (let [db @conn
+        tree (d/pull db
+                     '[[:db/id :as :eid]
+                       {:subsegments 1}]
+                     eid)]
+    (if (= eid 1)
+      (create-segment! recorder {:db db})
+      (create-subsegment! recorder {:db db :eid eid}))
+    (when-let [subsegments (seq (:subsegments tree))]
+      (doseq [{:keys [eid]} subsegments]
+        (send-trace! (assoc arg-map :eid eid))))
+    (if (= eid 1)
+      (.endSegment recorder)
+      (.endSubsegment recorder))))
+
+(defrecord AEntity [eid conn ^Clock clock ^AWSXRayRecorder recorder]
   ;; Implement AutoCloseable by default so that this can work with 'open-with'
   java.lang.AutoCloseable
-  (close [_]
+  (close [entity]
     (protocol/-close! entity))
 
   protocol/IAutoCloseable
-  (-close! [_]
-    (protocol/-close! entity))
+  (-close! [{:as entity :keys [eid conn]}]
+    (let [current-timestamp (current-timestamp-seconds clock)]
+      (d/transact! conn [{:db/id eid :end-at current-timestamp}]))
+    (when (= 1 eid)
+      (send-trace! entity))
+    nil)
 
   protocol/IEntity
-  (-set-exception! [this ex]
-    (when entity
-      (doto entity
-        (.addException ex)
-        (.setError true)))
-    this)
-
-  (-set-annotation! [this m]
-    (when entity
-      (doseq [[k v] m]
-        (let [^String key (csk/->snake_case_string k)]
-          (cond
-            (boolean? v) (.putAnnotation entity key ^Boolean v)
-            (number? v)  (.putAnnotation entity key ^Number v)
-            :else        (.putAnnotation entity key (str v))))))
-    this)
-
-  (-set-metadata! [this m]
-    (when entity
-      (doseq [[k v] m]
-        (.putMetadata entity (csk/->snake_case_string k) (str v))))
-    this))
+  (-set-exception! [{:as entity :keys [eid conn]} ex]
+    (d/transact! conn [{:db/id eid :exception ex}])
+    entity)
+
+  (-set-annotation! [{:as entity :keys [eid conn]} m]
+    (let [annotations (sanitize-keys m)]
+      (d/transact! conn [{:db/id eid :annotations annotations}])
+      entity))
+
+  (-set-metadata! [{:as entity :keys [eid conn]} m]
+    (let [metadata (sanitize-keys m)]
+      (d/transact! conn [{:db/id eid :metadata metadata}])
+      entity)))
 
 (extend-protocol protocol/IEntityProvider
   AWSXRayRecorder
-  (-start! [^AWSXRayRecorder recorder arg-map]
-    (let [^String name (get arg-map :name)
-          ^String trace-id (get arg-map :trace-id)
-          ^String parent-id (get arg-map :parent-id)]
-      (-> (if trace-id
-            (.beginSegment recorder name (TraceID/fromString trace-id) parent-id)
-            (.beginSegment recorder name))
-        (->AEntity))))
+  (-start! [^AWSXRayRecorder recorder {:as arg-map
+                                       :keys [clock]
+                                       :or {clock default-clock}}]
+    (let [conn (d/create-conn schema)
+          current-timestamp (current-timestamp-seconds clock)
+          segment (merge (select-keys arg-map [:name :trace-id :parent-id])
```

Comment:
A root span would be missing a parent-id and subsequently results in an error because you can't transact `nil` values.

Reply:
Got it. I'll deal with it later.
https://github.com/hden/aws-xray-sdk-clj/issues/19